use stores of the correct size to set discriminants

the8472 opened the pull request:
Resolves an old HACK /FIXME.

Note that I haven't worked much with codegen so I'm not sure if I'm using the functions correctly and I was surprised seeing out-of-range values being fed into `const_uint_big` but apparently they're wrapped implicitly? By making it explicit we can pass in-range values instead.

the8472 committed:
use stores of the correct size to set discriminants

the8472 commented:
CC @adwinwhite

RalfJung commented:
If this changes what codegen does, it should come with ui tests and/or codegen tests ensuring we don't regress this in the future.

bors commented:
:umbrella: The latest upstream changes (presumably #132581) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

estebank reviewed (approved):
r=me after rebase + test

the8472 committed:
add tests for niches in pointers

RalfJung commented on compiler/rustc_codegen_ssa/src/mir/place.rs:
```diff
@@ -387,13 +388,17 @@ impl<'a, 'tcx, V: CodegenObject> PlaceRef<'tcx, V> {
+                        bug!("expected a scalar placeref for the niche");
                     };
+                    let max = niche.layout.size.unsigned_int_max();
+                    let wrapped = niche_value & max;
```
```suggestion
                    // We used u128 arithmetic above, so we have to get this back in-bounds.
                    let wrapped = niche_value & max;
```

the8472 replied:
Added a more detailed comment.

the8472 commented:
I was waffling between making the test a build-pass (just checking codegen no longer ices) or run-pass. I moved to the latter and added some more asserts.

@bors r=RalfJung,estebank

bors commented:
:pushpin: Commit aec5b535be9bf1d596d33a2d8f6d57a7069537db has been approved by `RalfJung,estebank`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=RalfJung,estebank aec5b535be9bf1d596d33a2d8f6d57a7069537db -->
<!-- homu: {"type":"Approved","sha":"aec5b535be9bf1d596d33a2d8f6d57a7069537db","approver":"RalfJung,estebank","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

the8472 commented:
@bors r=RalfJung,estebank rollup

bors commented:
:pushpin: Commit cca2e96345a376b5e680d0962da3745bba6583b7 has been approved by `RalfJung,estebank`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=RalfJung,estebank cca2e96345a376b5e680d0962da3745bba6583b7 -->
<!-- homu: {"type":"Approved","sha":"cca2e96345a376b5e680d0962da3745bba6583b7","approver":"RalfJung,estebank","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

the8472 commented:
> I was waffling between...

And then I forgot to commit the changes 😔

@bors r=RalfJung,estebank

bors commented:
:pushpin: Commit b8d4b75844c6cd655dab67203fa5cc1d09bf6dca has been approved by `RalfJung,estebank`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=RalfJung,estebank b8d4b75844c6cd655dab67203fa5cc1d09bf6dca -->
<!-- homu: {"type":"Approved","sha":"b8d4b75844c6cd655dab67203fa5cc1d09bf6dca","approver":"RalfJung,estebank","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

RalfJung commented on compiler/rustc_codegen_ssa/src/mir/place.rs:
```diff
@@ -387,13 +388,21 @@ impl<'a, 'tcx, V: CodegenObject> PlaceRef<'tcx, V> {
                     };
+                    let max = niche.layout.size.unsigned_int_max();
+                    // Niche ranges that wrap around may be represented as out-of-range values
+                    // e.g. niche at 0u8 may end up being u8::MAX + 1.
```
I don't think that's true? AFAIK our niche ranges must always be inbounds of the scalar type they refer to. However, above you are doing `u128` arithmetic so even if the niche range endpoints are inbounds, your `niche_value` is not.

the8472 replied:
niche_start is inbounds, but niche_start + niche_value not. And this was passed as such to the backend, i.e. I have seen 256 passed to llvm for a repr(i8) enum via `const_uint_big`.

RalfJung replied:
`niche_start + niche_value` ought to be computed wrapping on the type of the niche field. But the code here takes a shortcut, but doing `u128`  arithmetic instead. *That* is why it can go out-of-bounds.

the8472 replied:
> niche_start + niche_value ought to be computed wrapping on the type of the niche field

If that were the case then it would have been buggy before. Note that I did not change the calculation and the out of range value was passed to the backend. This worked for integers. What I'm fixing is not that case but that llvm didn't like them for pointers.

RalfJung replied:
> And the result was passed as such to the backend, i.e. I have seen 256 passed to llvm for a repr(i8) enum via const_uint_big.

If that worked before, that means LLVM has been doing the truncation for integers. It just didn't like doing them for pointers, I guess.

RalfJung replied:
What your PR implements is exactly "do the arithmetic wrapping on the size of the niche field", though, so I don't understand what you are disagreeing with.

the8472 replied:
Yes, that's also my assumption. Which is why my comment is trying to explain that we do this to cover the pointer case which isn't handled by the backend (but for other cases they do handle it).

RalfJung replied:
Your comment says that niche values can go out-of-bounds.  That's just not correct. The correct niche value is computed by wrapping arithmetic on the size of the niche field and that cannot go out-of-bounds.

It'd be more clear to make this `let niche_value = (niche_value as u128).wrapping_add(niche_start) & max;`, so that the wrapping arithmetic is done in a single proper step.

the8472 replied:
What I don't understand is why you think my comment is incorrect.

We do calculate out of bound values. The backend accepts out of bound values in the common case. So we could just pass them through, as we did before. Ranges that wrap around resulting in out of bounds values is exactly what happens.

The only reason to do the wrapping ourselves is the special case that's currently only reachable via unstable attributes.

the8472 replied:
This calculates the niche value we tell the backend to set. Not the value we use internally to represent the niche.

The value we tell the backend can go out of bounds due to truncation in the backend.
I could have special-cased this to just do the wrapping outselves for pointers. But feels silly. Since they're equivalent I do it for integers too. But that's not a necessary property of the code.

the8472 replied:
Would this be better?

> The calculated niche_value may end up being out of bounds. This would be fine for integer types since such integer constants do get truncated by the backend to fit the type. But to also cover pointer types we always do it ourselves.

RalfJung replied:
I think your comment is misleading. The reason the value can be out-of-bounds is that it wasn't computed the right way. Or rather, it wasn't yet fully computed. The comment occurs half-way through the computation.

RalfJung replied:
I don't see why we would talk about the backend truncation behavior... it's a bit unclean to rely on that anyway (reasonably a backend could complain if we build an `i8` literal of value 256), and we are not relying on it any more with this PR.

RalfJung commented:
@bors r-

That is a substantially different comment than what I suggested, so we have to figure out which comment is correct here.

RalfJung commented on compiler/rustc_codegen_ssa/src/mir/place.rs:
```diff
@@ -387,13 +388,21 @@ impl<'a, 'tcx, V: CodegenObject> PlaceRef<'tcx, V> {
                     let niche_llty = bx.cx().immediate_backend_type(niche.layout);
                     let niche_value = variant_index.as_u32() - niche_variants.start().as_u32();
                     let niche_value = (niche_value as u128).wrapping_add(niche_start);
```
```suggestion
                    // We are supposed to compute `niche_value.wrapping_add(niche_start)` wrapping
                    // around the `niche`'s type. Since we know that type to be unsigned,
                    // the easiest way to do that is to do wrapping arithmetic on `u128` and then
                    // masking off any extra bits that occur because we did the arithmetic with too many bits.
                    let niche_value = (niche_value as u128).wrapping_add(niche_start);
                    let niche_value = niche_value & niche.layout.size.unsigned_int_max();
```

RalfJung replied:
Turns out it can be signed, but the logic should be right even for that case.

RalfJung commented:
Thanks!

@bors r+

bors commented:
:pushpin: Commit 26d7b5da996ec8935c33638f1d7c29f58c96f8f0 has been approved by `RalfJung`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=RalfJung 26d7b5da996ec8935c33638f1d7c29f58c96f8f0 -->
<!-- homu: {"type":"Approved","sha":"26d7b5da996ec8935c33638f1d7c29f58c96f8f0","approver":"RalfJung","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
